### DISCUSSION TITLE
How to use in combination of Qt Test?

### DISCUSSION THREAD

**Discussion by LuisAyuso** [faaxm/spix#25] [General]
Hi there,

First of all, terrific job with this library, I can not believe that something like this is not part of the standard tooling. 

I have been experimenting with using this tool in combination with a c++ backend, using Gtest.
I ran into all kinds of errors because of the threads spawned outside of Qt control  (I presume GTest runs fixtures in spawned threads).
For my use case, I need some testing framework to report in xml and such. Therefore my own main function is not an option.

Would be possible to combine the Spix test server with a Qt standard test object? 
Is it possible to tell GTest to drop extra threading? 

Cheers

**Comment by faaxm** [comment]
Hi!

Happy to hear you like Spix :-)
Spix itself already spawns a standard thread in which the actual test is executed, so that it runs independently from the UI. However, in my experience, this should not cause any issues with Qt.

Did you already have a look at the GTest example? It sounds to me it does exactly what you are looking for:
https://github.com/faaxm/spix/tree/master/examples/GTest

**Comment by LuisAyuso** [comment]
I could not manage to get it working correctly with GTest, the use case is as follows:

- There is a main function that initializes Qt, Spix and registers my qml MetaTypes.
- There is a test server just like one in your example.
- The test server launches the TEST macro fixtures
- I construct a QQuickView object in the fixture to test some branch of my QML tree, 
It complains about parent thread when constructing some objects (I presume QML Timer object is related to this) 

When I go one step ahead and initialize my complete c++ backend in the test fixture, the logs multiply and the application does 
not correctly initialize.

**Reply by faaxm** [reply]
Spix was originally designed to test user interaction in the full application, less with a unit-test like setup in mind (i.e. loading different subsets of the QML per test). As such, it assumes that a single UI has been already constructed in the main thread and then runs the tests on that UI from a separate thread. Constructing the UI in the fixture won't work, as you have found. You could try the following workaround, though I haven't tried it myself:

While the test itself (i.e. the test commands for spix) are executed in a background thread, the actual commands are then executed in the main thread (posting events, ...). This means you could try the following:
* register a "genericCommandHandler" on the TestServer (https://github.com/faaxm/spix/blob/master/lib/src/TestServer.cpp#L49) in which you load a new QML branch, depending on the `payload` string (this will be executed in the main thread)
* in your test fixture, call the "generic command" to load the QML branch you would like to test (https://github.com/faaxm/spix/blob/master/lib/src/TestServer.cpp#L83). Maybe you can just use the name of the qml file as payload?

Let me know if this works.

**Reply by Laxx18** [reply]
Hi, 

has someone solved this issue and is willing to paste example code? That would be really great, as I'm also having this issue and have no idea how to solve it.

Best Regards
Lax

**Reply by faaxm** [reply]
I presume you already had a look at the example here?
https://github.com/faaxm/spix/blob/master/examples/GTest

Does that example run on your machine?

**Reply by Laxx18** [reply]
No, because we are using QTest and it does work totally different as GTest.
